src/com/github/Kraken3/AFKPGC/BotDetector.java

```diff
@@ -0,0 +1,306 @@
+							// Its movement looks botlike.
+							AFKPGC.debug("Player ", curSuspect.getUUID(), " looks like a bot (bounds test ", 
+									truebot, "): ", bounds);
+
+							// Now test surrounding area.
+							Location point = curSuspect.getLocation();
+
+							LagScanner ls = new LagScanner(point, scanRadius, null);
+							ls.run(); // TODO: move this and ban results to thread.
+							if (ls.isLagSource()) {
+								thisRoundSuspect = curSuspect;
+								Date currentDate = new Date();
+								long bantime = (long) (frequency / currentTPS * 1000);
+								/* Because the ban time is based on real time and the next run
+								 * of this method is based on the tick, the ban time needs to be
+								 * adjusted to the tick. The long form of this would be:
+								 * (20/currentTPS) * (frequency/20) * 1000
+								 * The time is in ms */
+								BanEntry leBan = banList.addBan(curSuspect.getName(),
+										"You were suspected to cause lag and banned for "
+												+ (bantime / 1000) + " seconds",
+												new Date(currentDate.getTime() + bantime), null);
+								Player p = Bukkit.getPlayer(curSuspect.getUUID());
+								if (p != null) {
+					   				p.kickPlayer(leBan.getReason());
+								}
+								// ban the player briefly and skip the other suspects.
+								AFKPGC.debug("Player ", curSuspect.getUUID(), " (", curSuspect.getName(),
+										") exceeded ban threshold with ", ls.getLagCompute(), " banned for",
+										bantime / 1000, " milliseconds");
+								break;
+							} else {
+								AFKPGC.debug("Player ", curSuspect.getUUID(), " (", curSuspect.getName(),
+										") cleared via insufficient lagsources [", ls.getLagCompute(), "]");
+							}
+						} else {
+							AFKPGC.debug("Player ", curSuspect.getUUID(), " (", curSuspect.getName(),
+									") cleared via bounding box [", truebot, "]");
+						}
+						// It passed the truebot(tm) detection -- for now. Give a temporary reprieve.
+						reprieve.put(curSuspect.getUUID(), maxReprieve);
+						AFKPGC.debug("Player ", curSuspect.getUUID(), " unlikely bot, given reprieve (bounds test ",
+								truebot, "): ", bounds);
+					} else {
+						// else not enough info yet. Pass.
+						AFKPGC.debug("Player ", curSuspect.getUUID(), " Not enough bounds data, skip for now");
+					}
+				}
+			} else {
+				AFKPGC.debug("No suspects this round.");
+			}
+
+			if (lastRoundSuspect != null) {
+				// TODO: This is starting to stink of synchronization issues. We shouldn't need
+				//       to check immune so often within the same compute round.
+				if (!AFKPGC.immuneAccounts.contains(lastRoundSuspect.getUUID())) {
+					if (currentTPS - lastRoundTPS > criticalTPSChange) {
+						/* This can be relatively sensitive, because it will only ban players for 
+						 * a longer period of time, if it catches them twice here */
+						Date currentDate = new Date();
+						if (suspectedBotters.contains(lastRoundSuspect.getName())) {

```

Comment:
Ah, I just had in mind that lastRoundSuspect was a Player, from when I was writing it and didn't see that you changed it, my bad.